src/Build/Instance/ProjectItemInstance.cs

```diff
@@ -929,7 +929,12 @@ public ICollection MetadataNames
             {
                 get
                 {
-                    List<string> names = new List<string>((List<string>)CustomMetadataNames);
+                    List<string> names = new List<string>(capacity: MetadataCount);
```

Comment:
Note: this is only really an improvement if MetadataCount is very fast—which it currently might not be.

Also, is there some convenient method like:
return MetadataCollection.Count.Union(FileUtilities.ItemSpecModifiers.All);
?

`All` is an array, so I don't think so, but thought I'd check.

Reply:
`MetadataCount` has been updated to be much faster. Please take another look and see if you feel this needs further work.

Reply:
I said Count but meant Keys.

return MetadataCollection.Keys.Concat(FileUtilities.ItemSpecModifiers.All);
?

Reply:
That's possible, though we'd need to convert the `IEnumerable<>` to an `ICollection` . Calling `ToList` or `ToArray` is going to result in potentially multiple redundant allocations, as the resulting collection size is not known from the enumerator produced by `Concat`.

By allocating a list up-front with the needed capacity, that overhead is avoided and no waste occurs in the backing arrays.

It would be possible to make this allocate one fewer object by using `string[]` directly, rather than `List<string>`.